### PR TITLE
Ref #406 -- Allow manual reorder in waiting list

### PR DESCRIPTION
**Pull request by danlls**
- Add model column 'priority' as sorting criteria
- Add a new page to reorder priority of entries which has not been
  assigned a voucher.

### CODE REVIEW

**Comment by danlls** [comment]
Not sure why build is failing on postgresql config only. Any insight?

**Comment by rixx** [comment]
When restarting the build of one of the failing scenarios, it errored out due to conflicting migrations – would you mind rebasing against the commit 86f1709 which is the last passing commit on the master branch?

Rebasing against a specific commit is done by checking out your feature branch and running `git rebase 86f1709`. – If you run into issues or don't feel comfortable with this, I could do it, too.

Afterwards you'll probably need to fix your migration to depend on the last migration in the current pretix line, which means changing the `dependencies` list in your migration to read `('pretixbase', '0071_auto_20170729_1616')`.

Afterwards, tests should be able to run again, and we can have a look at the postgres issue if it persists.

**Comment by danlls** [comment]
Weird, rebased and still failing 😢

**Comment by danlls** [comment]
Alright, thanks for your help!

File: src/pretix/base/migrations/0072_auto_20170803_1006.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def set_priority(app, schema_editor):
+    wles = app.get_model('pretixbase', 'WaitingListEntry')
+    for wle in wles.objects.all():
+        if wle.voucher:
+            wle.priority = None
+        else:
+            wle.priority = 10
+        wle.save()
```
Comment (michael-k) [inline src/pretix/base/migrations/0072_auto_20170803_1006.py]: Number of Database Queries: 2 * #WaitlingListEntry + #WaitingListEntryWithVoucher

`wles.objects.filter(voucher__isnull=True).update(priority=10)`

More details:
- `wle.voucher` leads to a db query (check `wle.voucher_id` instead or use `select_related`)
- `wle.priority = None` is the default, no need to call save (=> hint that `filter` should be used)
- `wle.save()` sends all data to the database; consider using `save`'s `update_fields` instead

File: src/pretix/base/models/waitinglist.py [hunk 2]
```diff
@@ -136,3 +142,10 @@ def send_voucher(self, quota_cache=None, user=None):
                 self.event,
                 locale=self.locale
             )
+
+    def save(self, *args, **kwargs):
+        if self.voucher is not None:
```
Comment (michael-k) [inline src/pretix/base/models/waitinglist.py]: Consider checking `self.voucher_id` instead.

File: src/pretix/base/migrations/0072_auto_20170803_1006.py (hunk 1 above)
Comment (danlls) [inline src/pretix/base/migrations/0072_auto_20170803_1006.py]: Thanks for the neat suggestions to reduce database hit count. I've learned much! 😄

**Comment by raphaelm** [comment]
I debugged the postgres failure: The test ``test_delete_single`` tries to delete an entry from the list. It just randomly grabs the first one. Due to your new default ``ordering`` and the internal ordering for postgres, this now accidentally happens to be the one list entry that already has a voucher and therefore can no longer be deleted. Just changing the respective line to ``wle = WaitingListEntry.objects.filter(voucher=None).first()`` in the test helps.


Otherwise, I see one open question here. You now have implemented a separate page that shows a form field for the priority of every entry on one page. This is very convenient if you want to change multiple ones and certainly a good idea *until* you have more than 30 entries and the list begins to paginate. Suddenly, it probably becomes inconvenient because (a) you need to know that you must first save, then switch to the first page and (b) the changes you made on the first page might change the order of everything such that the second page will *not* be what you expect, since you might see some of the ones you already changed again and miss some others completely.

I'm not sure what the good solution is here. The *easy* solution certainly is to switch back to a simple update view that just allows changing one entry at a time. Do you have other ideas how we could approach this?

**Comment by danlls** [comment]
I'm not sure if the following idea will work, but here it is:

Maybe instead of using `ModelFormSet`, we can use a normal `FormSet` to act as a proxy form.

1. Fields of formset initial value will be set to be as `wle.priority`
2. Any modification to each field will be saved/recorded using ajax or on pagination to ensure persistence when user go back and forth pages.
3. These modification won't directly modify the database, list ordering won't change upon pagination.
4. Once user actually clicked on `Save`, then copy all changes and apply it on the `WaitingListEntry` model.

**Comment by raphaelm** [comment]
Yep, I guess that would work, but it feels like more effort that it's worth.

**Pull request closed without merging**